### PR TITLE
Reduce usages of `com.google.common.collect` in favor of native Java functionality

### PR DESCRIPTION
**Pull request by basil**
A subset of #5344 that is hopefully non-controversial and safe to land in the short term because it avoids any changes that have a risk of regression.

The underlying motivation is to make the code easier to read by using a consistent paradigm as much as possible (namely, JDK classes rather than a mixture of JDK classes and Guava classes).

### CODE REVIEW

File: core/src/main/java/hudson/ExtensionFinder.java [hunk 1]
```diff
@@ -305,12 +305,14 @@ private void refreshExtensionAnnotations() {
...
+        private List<IndexItem<?, Object>> loadSezpozIndices(ClassLoader classLoader) {
             List<IndexItem<?,Object>> indices = new ArrayList<>();
             for (GuiceExtensionAnnotation<?> gea : extensionAnnotations.values()) {
-                Iterables.addAll(indices, Index.load(gea.annotationType, Object.class, classLoader));
+                for (IndexItem<?, Object> indexItem : Index.load(gea.annotationType, Object.class, classLoader)) {
+                    indices.add(indexItem);
+                }
             }
-            return ImmutableList.copyOf(indices);
+            return Collections.unmodifiableList(indices);
```
Comment (basil) [inline core/src/main/java/hudson/ExtensionFinder.java:315]: `indices` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/hudson/ExtensionFinder.java [hunk 2]
```diff
@@ -666,7 +668,7 @@ public synchronized ExtensionComponentSet refresh() {
 
             List<IndexItem<Extension,Object>> r = new ArrayList<>(old);
             r.addAll(delta);
-            indices = ImmutableList.copyOf(r);
+            indices = Collections.unmodifiableList(r);
```
Comment (basil) [inline core/src/main/java/hudson/ExtensionFinder.java:671]: `r` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 3]
```diff
@@ -702,10 +702,10 @@ public final void post(BuildListener listener) throws Exception {
                 post2(listener);
             } finally {
                 // update the culprit list
-                HashSet<String> r = new HashSet<>();
+                SortedSet<String> r = new TreeSet<>();
                 for (User u : getCulprits())
                     r.add(u.getId());
-                culprits = ImmutableSortedSet.copyOf(r);
+                culprits = Collections.unmodifiableSet(r);
```
Comment (basil) [inline core/src/main/java/hudson/model/AbstractBuild.java:708]: `r` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/hudson/model/DependencyGraph.java [hunk 4]
```diff
@@ -213,13 +212,13 @@ private DependencyGraph(boolean dummy) {
...
-            return ImmutableList.of();
+            return Collections.emptyList();
         } else {
-            ImmutableList.Builder<Dependency> builder = ImmutableList.builder();
+            List<Dependency> builder = new ArrayList<>();
             for (DependencyGroup dependencyGroup : v) {
                 builder.addAll(dependencyGroup.getGroup());
             }
-            return builder.build();
+            return Collections.unmodifiableList(builder);
```
Comment (basil) [inline core/src/main/java/hudson/model/DependencyGraph.java:221]: `builder` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/hudson/model/Fingerprint.java [hunk 5]
```diff
@@ -1177,13 +1178,13 @@ public synchronized boolean trim() throws IOException {
             for (TransientFingerprintFacetFactory fff : TransientFingerprintFacetFactory.all()) {
                 fff.createFor(this,transientFacets);
             }
-            this.transientFacets = ImmutableList.copyOf(transientFacets);
+            this.transientFacets = Collections.unmodifiableList(transientFacets);
```
Comment (basil) [inline core/src/main/java/hudson/model/Fingerprint.java:1181]: The local variable `transientFacets` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/hudson/model/queue/MappingWorksheet.java [hunk 6]
```diff
@@ -370,7 +370,7 @@ public MappingWorksheet(BuildableItem item, List<? extends ExecutorSlot> offers,
             if (ec.node==null)  continue;   // evict out of sync node
             executors.add(ec);
         }
-        this.executors = ImmutableList.copyOf(executors);
+        this.executors = Collections.unmodifiableList(executors);
```
Comment (basil) [inline core/src/main/java/hudson/model/queue/MappingWorksheet.java:373]: The local variable `executors` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/hudson/model/queue/MappingWorksheet.java [hunk 7]
```diff
@@ -387,7 +387,7 @@ public MappingWorksheet(BuildableItem item, List<? extends ExecutorSlot> offers,
         for (List<SubTask> group : m.values()) {
             works.add(new WorkChunk(group,works.size()));
         }
-        this.works = ImmutableList.copyOf(works);
+        this.works = Collections.unmodifiableList(works);
```
Comment (basil) [inline core/src/main/java/hudson/model/queue/MappingWorksheet.java:390]: The local variable `works` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/hudson/tasks/Fingerprinter.java [hunk 8]
```diff
@@ -461,7 +461,7 @@ public AbstractBuild getBuild() {
                 }
             }
 
-            m = ImmutableMap.copyOf(m);
+            m = Collections.unmodifiableMap(m);
```
Comment (basil) [inline core/src/main/java/hudson/tasks/Fingerprinter.java:464]: The local variable `m` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/jenkins/agents/WebSocketAgents.java [hunk 9]
```diff
@@ -87,11 +88,11 @@ public HttpResponse doIndex(StaplerRequest req, StaplerResponse rsp) throws IOEx
...
-        state.fireAfterProperties(ImmutableMap.of(
-            // TODO or just pass all request headers?
-            JnlpConnectionState.CLIENT_NAME_KEY, agent,
-            JnlpConnectionState.SECRET_KEY, secret
-        ));
+        // TODO or just pass all request headers?
+        Map<String, String> properties = new HashMap<>();
+        properties.put(JnlpConnectionState.CLIENT_NAME_KEY, agent);
+        properties.put(JnlpConnectionState.SECRET_KEY, secret);
+        state.fireAfterProperties(Collections.unmodifiableMap(properties));
```
Comment (basil) [inline core/src/main/java/jenkins/agents/WebSocketAgents.java:95]: `properties` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 10]
```diff
@@ -2571,11 +2569,12 @@ private File expandVariablesForDirectory(String base, Item item) {
...
-                "JENKINS_HOME", Jenkins.get().getRootDir().getPath(),
-                "ITEM_ROOTDIR", itemRootDir,
-                "ITEM_FULLNAME", itemFullName,   // legacy, deprecated
-                "ITEM_FULL_NAME", itemFullName.replace(':','$'))); // safe, see JENKINS-12251
+        Map<String, String> properties = new HashMap<>();
+        properties.put("JENKINS_HOME", Jenkins.get().getRootDir().getPath());
+        properties.put("ITEM_ROOTDIR", itemRootDir);
+        properties.put("ITEM_FULLNAME", itemFullName); // legacy, deprecated
+        properties.put("ITEM_FULL_NAME", itemFullName.replace(':','$')); // safe, see JENKINS-12251
+        return Util.replaceMacro(base, Collections.unmodifiableMap(properties));
```
Comment (basil) [inline core/src/main/java/jenkins/model/Jenkins.java:2577]: `properties` was declared within this method and goes out of scope at the end of the method, so nothing else can modify it. We do not need to copy it.

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by StefanSpieker — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
All looks good. It is hard to estimate the potential downsides though. Performance should not be a problem at least, native Java implementations are commonly better on benchmarks for modern Java versions

**Comment by saper** [comment]
Could this be causing https://issues.jenkins.io/browse/JENKINS-65611 that breaks 2.292 for some people?

**Pull request merged**